### ISSUE TITLE
Adding X-Ray tracing and permissions to existing function/role fails

### ISSUE THREAD

**Issue by lafiosca** [awslabs/serverless-application-model#225]
_Overview_

If you have created a Lambda function and role with SAM from an `AWS::Serverless::Function` resource without X-Ray tracing, and you later try to add the "XRayWriteOnlyAccess" managed policy and enable tracing at the same time, the CloudFormation deployment fails with a ResourceStatusReason of "The provided execution role does not have permissions to call PutTraceSegments on XRAY".

_Sample Files_

**sam-template1.json**
```
{
    "AWSTemplateFormatVersion": "2010-09-09",
    "Transform": "AWS::Serverless-2016-10-31",
    "Description": "Xray test",
    "Resources": {
        "TestFunction": {
            "Type": "AWS::Serverless::Function",
            "Properties": {
                "FunctionName": { "Fn::Sub": "${AWS::StackName}-Test" },
                "Handler": "index.handler",
                "Runtime": "nodejs6.10",
                "CodeUri": "test.zip",
                "MemorySize": 128,
                "Timeout": 10,
                "Description": "Test function",
                "Policies": [
                    "AmazonDynamoDBReadOnlyAccess"
                ]
            }
        }
    }
}
```

**deploy1.sh**
```
#!/bin/bash
aws cloudformation package --template-file sam-template1.json --output-template-file sam-output1.yml --s3-bucket your-artifact-bucket --s3-prefix test
aws cloudformation deploy --template-file sam-output1.yml --stack-name XrayTest --capabilities CAPABILITY_IAM
```

**sam-template2.json**
```
{
    "AWSTemplateFormatVersion": "2010-09-09",
    "Transform": "AWS::Serverless-2016-10-31",
    "Description": "Xray test",
    "Resources": {
        "TestFunction": {
            "Type": "AWS::Serverless::Function",
            "Properties": {
                "FunctionName": { "Fn::Sub": "${AWS::StackName}-Test" },
                "Handler": "index.handler",
                "Runtime": "nodejs6.10",
                "CodeUri": "test.zip",
                "MemorySize": 128,
                "Timeout": 10,
                "Description": "Test function",
                "Policies": [
                    "AmazonDynamoDBReadOnlyAccess",
                    "AWSXrayWriteOnlyAccess"
                ],
                "Tracing": "Active"
            }
        }
    }
}
```

**deploy2.sh**
```
#!/bin/bash
aws cloudformation package --template-file sam-template2.json --output-template-file sam-output2.yml --s3-bucket your-artifact-bucket --s3-prefix test
aws cloudformation deploy --template-file sam-output2.yml --stack-name XrayTest --capabilities CAPABILITY_IAM
```

_Steps to Reproduce_

1. Copy these files into a directory
2. Replace `your-artifact-bucket` in the deploy scripts with an S3 bucket you have write access to
3. Copy a test Lambda package zip file into the same directory as `test.zip`
4. Run `deploy1.sh` and wait for the `XrayTest` stack to be created successfully
5. Run `deploy2.sh` and wait for the CloudFormation deployment on the `XrayTest` stack to fail

_Workaround_

If you perform an intermediate deployment to add the "AWSXrayWriteOnlyAccess" policy without enabling X-Ray tracing, it will work. Then you can perform another deployment to enable tracing.

**Comment by lafiosca** [comment]
I believe this is a bug, and I am naively guessing that it has something to do with the timing of the resource updates.

**Comment by jfuss** [comment]
Hey @lafiosca,

This is very similar, if not the same as #217. Have you looked at this issue?

This is (to my knowledge) very much a timing issue and I think #217 does a fair job explaining what is going on.

Does this happen every time? Have you tried re-deploying the template a second time? Wondering if it happens every time?

**Comment by lafiosca** [comment]
Thanks, @jfuss. I didn't see #217 because I was only searching open issues. I agree that it appears to be the same problem, although I disagree with closing the issue while it still exists, even if the root cause is CloudFormation.

I believe it happens every time. In my example, running `deploy2.sh` multiple consecutive times does not fix the problem because the stack rolls back on failure, un-adding the role's new permission.

Regarding your comments from #217, I'm curious why you discounted Custom Resources as a stop-gap measure. I am not extremely familiar with them, but I would guess you could use one to implement a blunt delay like you suggested. I imagine the SAM transform would yield a Lambda function resource which depends on the custom resource, which further depends on the Lambda execution role resource. The custom resource would be backed by a Lambda which just waited a while (carefully to avoid tripping the max execution time). I don't really love this solution, but if it worked it would perhaps be better than having to split our work into multiple commits/deployments.

**Comment by lafiosca** [comment]
(On a side note, I'll mention that I believe this same delayed role bug may affect the web console as well. Our first experience with X-Ray was attempting to enable X-Ray for an existing Lambda manually by checking the box and hitting save, and the console gave a permissions error on the first try but worked on subsequent tries.)

**Comment by jfuss** [comment]
@lafiosca I still think closing the issue is correct. If someone runs into an issue and doesn't look to see if there has been someone else with a similar issue that was closed and always cuts new issues, than open issues become very noisy with things that have been covered. However, I do get your point you are raising.

I did not try to discount them completely, I still don't completely understand how it would be helpful in this situation. This is more due to my lack of knowledge in Custom Resources than them not being helpful. What does concern me here, is that we could be supporting this 'delay' resource for much more than SAM. Once we release a Custom Resource in SAM, it will become available to all if they know what the resource schema is, right? If that is correct, this becomes a much bigger discussion with the team. 

More and more I think about this, the more and more I think we should be taking a step back and look at this from fresh eyes (instead of reading other issues and saying ya that seems like the same thing). What I would like to do is go a little deeper into the issue and see if we can really narrow down what is going on here. I can think of two potential causes that would result in this issue. The team wont be able to get to this immediately with everything else going on though. I will work to get this prioritize in some fashion, so we can do a deeper investigation. That will also help us in the Custom Resource discussion, if we really need it. In the meantime, the best we can offer is the multiple deployment paths to get around this.
